### DISCUSSION TITLE
Unable to get sensor data from recorded file using world.get_actors()

### DISCUSSION THREAD

**Discussion by zl3466** [carla-simulator/carla#5525] [Q&A]
Hi all, I am currently trying to get existing sensors from a recorded .log file with the function world.get_actors(). However, it appears that the function only returns the landmarks (type = 'traffic.*') and vehicles (type = 'vehicle.*'), but not sensors (type = 'sensors.*'). Meanwhile, running show_recorder_file_info.py upon the .log file gives me proper records of creations of the sensors.

what I did in terms of coding is simply adding
world = client.get_world()
print(world.get_actors().filter('sensors.*'))
directly after the client.replay_file function call in the start_replaying.py file from the API example folder.

A similar problem has been raised [here](https://github.com/carla-simulator/carla/issues/3864) and claimed to be a bug, but the follow up comment says that the bug has been fixed. Yet I am still unable to retrieve sensors with world.get_actors() from recordings. 

I am not sure if this matters, but the recording was generated with sync mode = True during the simulation, as I was visualizing multiple sensors (showing multiple camera and lidar views using cv2.imshow()) in the process. [This thread](https://github.com/carla-simulator/carla/issues/4144) says it should be fine as long as I am running the replay in the same map as I performed the simulation, which I did. 

My ultimate goal is to reproduce the multi-sensor visualization from recording .log files, which requires the extraction of spawned sensors, to the data of which I can listen(), callback, process, and visualize.

here's the partial code from my modified start_replaying.py. the get_actors() action is at the very bottom.

    try:

        client = carla.Client(args.host, args.port)
        client.set_timeout(60.0)

        # set the time factor for the replayer
        client.set_replayer_time_factor(args.time_factor)

        # set to ignore the hero vehicles or not
        client.set_replayer_ignore_hero(args.ignore_hero)

        # replay the session
        print(client.replay_file(args.recorder_filename, args.start, args.duration, args.camera, args.spawn_sensors))
        
        world = client.get_world()
        print(world.get_actors().filter('vehicle.*'))

    finally:
        pass

Plzzz let me know if anyone has any idea how to properly extract sensor data from recording files. Thanks!!!

**Comment by Morphlng** [comment]
Hello! I'm wondering if you have solved this problem?

We are trying to collect data from a series of replay files. `get_actor(id)` will return the sensor I want, but the problem to me is that only "RGB" camera can register a callback and work, the CollisionSensor and LaneInvasionSensor in my case, doesn't call the callback function I registered through `listen` at all.

I find that the replay file itself record the event of collision, however there are no events for lane_invasion. I cannot get that data without a functioning callback. I have tried Carla 0.9.13 and Carla 0.9.14, both doesn't work.

**Reply by LarryYang1** [reply]
@Morphlng 

Hello! I want to do the same thing as you did. I firstly make a record through a simulation with 6 cameras, then I want to replay the recording, and handle 6 cameras' input for each frames. Should I create new 6 cameras to register callbacks, or use the existing 6 cameras to register callbacks? I noticed during replay, there already have 6 cameras, but all their attributes data are wrong... Could you share some sample code to show how you make it work? Thanks in advance!

**Reply by Morphlng** [reply]
Hi @LarryYang1 ! I don't remember where I have seen this, but the mechanism of Carla's replay is "teleporting" the actors frame by frame, to simulate its "moving". So some of the data (e.g. `vehicle.get_velocity()`) and sensor data (e.g. LaneInvasion) is not able to retreive. If you only want RGB camera data, then I recommend you to setup some new sensor after you `client.replay_file`.

I don't have some ready-to-go scripts, but the script should be something like this:

1. `client.load_world` manually to which the record happened. (Recommended)
2. `world.apply_settings` to switch to `synchronous_mode`. (It is also recommended to record in sync mode with fixed delta seconds)
3. `client.replay_file` to load the record. (I once wrote a [replay_parser](https://github.com/Morphlng/Carla-Interactive-Script/blob/main/src/replay_parser.py) for Carla, that can get all vehicle's id beforehand. This should work under Carla 0.9.13)
4. `world.tick` until all vehicles are spawned.
5. Setup sensors at this point.
6. Finish your data collection logic.

**Reply by LarryYang1** [reply]
@Morphlng, Thanks for your quick response, I can't get camera callback data with **new** sensors, but it seems I can get camera callback if I register those callbacks to **existing** camera sensors, after replay_file, then I can get sensor data from existing IMU, GNSS, camera, but just no callback data for speedometer. I noticed that the attributes for the existing cameras are wrong, anyway, I will check the input data from existing cameras and see whether they are as expected or not...
